src/3d/lights/qgslightsource.h

```diff
@@ -0,0 +1,94 @@
+/***************************************************************************
+                          qgslightsource.h
+                          ---------------
+    begin                : April 2022
+    copyright            : (C) 2022 by Nyall Dawson
+    email                : nyall dot dawson at gmail dot com
+***************************************************************************/
+
+/***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#ifndef QGSLIGHTSOURCE_H
+#define QGSLIGHTSOURCE_H
+
+#include "qgis_3d.h"
+#include "qgis.h"
+#include "qgis_sip.h"
+#include "qgsreadwritecontext.h"
+
+#include <QList>
+
+class Qgs3DMapSettings;
+class QgsProject;
+class QDomElement;
+class QDomDocument;
+
+#ifndef SIP_RUN
+namespace Qt3DCore
+{
+  class QEntity;
+}
+#endif
+
+/**
+ * \ingroup 3d
+ * \brief Base class for light sources in 3d scenes.
+ *
+ * \since QGIS 3.26
+ */
+class _3D_EXPORT QgsLightSource SIP_ABSTRACT
+{
+
+  public:
+
+    virtual ~QgsLightSource();
+
+    /**
+     * Returns the light source type.
+     */
+    virtual Qgis::LightSourceType type() const = 0;
+
+    /**
+     * Returns a copy of the light source.
+     */
+    virtual QgsLightSource *clone() const = 0 SIP_FACTORY;
+
+    /**
+     * Creates entities representing the light source.
+     */
+    virtual QList< Qt3DCore::QEntity * > createEntities( const Qgs3DMapSettings &map, Qt3DCore::QEntity *parent ) const = 0 SIP_SKIP;
```

Comment:
To keep things tidier, I would slightly prefer to have this return just a single QEntity, and for the case where the light source origin should be shown, there could be another virtual function returning another QEntity...

Reply:
I have a follow up feature which is a "point lights from layer" source. That one creates a light at each point feature, so there'll be multiple entities created. Does your concern still apply?

Reply:
I see... But also with "point lights from layer", all the lights could be parented to a single QEntity which would be then returned. Maybe it could simplify the management of entities in the scene, I don't know :-) I don't have strong opinion, so if you are happy with createEntities() then feel free to merge

Reply:
> , all the lights could be parented to a single QEntity which would be then returned

Great point👍

Reply:
Done